examples/3rdparty/python/BUILD

```diff
@@ -4,6 +4,28 @@
 # see/edit requirements.txt in this directory to change deps.
 python_requirements()
 
+# NB: google-cloud-bigquery has the following prelude upon import:
+#
+# from pkg_resources import get_distribution
+# __version__ == get_distribution("google-cloud-bigquery").version
+#
+# which will fail at import time with pkg_resources.DistributionNotFound.
+# The argument `prepend_to_pythonpath=True` will work around this particular error by ensuring the
+# requirements are within a pex file located at the at the front of the PEX_PATH when executing
+# pytest. Note that even though this works, it will still show the following warning:
```

Comment:
Is this actually pytest specific? The test you added doesn't use pytest at all, it just tests `./pants run` on a `python_binary`. Maybe remove the references to `pytest` here?

Reply:
In this case, it is specific to the way we construct the pytest pex. The error fixed by `prepend_to_pythonpath=True` does *not* show up with `PythonRun` -- see https://github.com/pantsbuild/pants/blob/e42f69d634dd78eed2750914038cb6e702a82a55/tests/python/pants_test/backend/python/tasks/test_bigquery_python_run_integration.py#L7-L8

The import issue is specifically for these packages in the context of pants's `PytestRun` task -- I will make this more clear instead of just saying "pytest".

Reply:
Aha - in that case, the added test should probably run `./pants test` rather than `./pants run`? Presumably the unit test passed before this change, as well as after?

Reply:
Even still, I would remove `when executing pytest`. This change should apply anytime this library is used in a PEX.

Reply:
> Aha - in that case, the added test should probably run ./pants test rather than ./pants run?

The invocation of `./pants run` in the integration test is testing a separate issue, which is also fixed in this PR. The use of `SelectInterpreterClientMixin` here is intended so that the interpreter pinning from `PytestRun` in #7563 could be used in `PythonRun`, which fixes this separate interpreter selection failure in `PythonRun`.

The reason this change couples those two fixes is that the interpreter selection failure in `PythonRun` that this change fixes doesn't occur with any random package I could find on pypi when testing -- specifically `google-cloud-bigquery` (all versions I tried) happened to also be the repro case for both the `prepend_to_pythonpath=True` fix for `PytestRun`, as well as this separate interpreter selection fix for `PythonRun`.

I don't at all understand why this `google-cloud-bigquery` package (several other related google cloud packages as well, unfortunately) would be causing the interpreter selection failure specifically, I really cannot see what they are doing wrong to cause that issue and it may also repro for other packages I haven't tried. But it is ~clear (from the pex warning message which says exactly this) that they should be putting `setuptools` in their `install_requires`, and they're not. I don't know how bad that is or whether it would be relevant to the `PythonRun` interpreter selection issue that this change also fixes. I looked for some time for a way to somehow invoke pex or create a pex differently to make sure different versions of setuptools were available, and this was the only workaround I could find.

One of the remaining unknowns to me is where this broke, since our monorepo has this dependency running just fine in pants 1.15.x. I was not able to identify any behavior which differed from pex `1.6.5` to `1.6.6` (although I am less familiar with pex than I could be), so something changed in pants. I would love to fix that, but haven't figured it out yet.

Perhaps a TODO today is to try to bisect the change in behavior before trying to merge this.
> Presumably the unit test passed before this change, as well as after?

The statement `from google.cloud import bigquery` would fail if run in a python test before this change with `pkg_resources.DistributionNotFound` -- so this integration test would have failed, but failed at import time, before running tests. That's what `prepend_to_pythonpath=True` works around. The integration test then does `./pants run` in order to test the separate `PythonRun` issue this change also fixes.
> This change should apply anytime this library is used in a PEX.

Well, the `prepend_to_pythonpath` fix actually doesn't appear to be necessary outside of `PytestRun`, which is strange since we're using `self.create_pex()` in `PytestPrep` in a very normal way. It's not yet clear to me why this is the case.